### PR TITLE
Update buildchain and CI for operator

### PR DESCRIPTION
**Pull request by slaperche-scality**
**Component**:

ci, buildchain

**Context**: 

In order to deploy the operator, we need to be able to build it and add it in the ISO.

**Summary**:

Add new dependencies in the CI to build the operator (`go`, `operator-sdk` and `dep`).
Update the buildchain with new tasks:
- to build the operator image and add it to the ISO
- to lint the Go code (and check that the generated files are up-to-date)
- to help formatting Go code (using `gofmt`) and generating boilerplate code (using `operator-sdk generate`).

**Acceptance criteria**: 

When building the ISO, you should have the operator image in the build tree (check with `ls _build/root/images/storage-operator/*`)
If you're familiar with Go, you can also edit `storage-operator/pkg/apis/storage/v1alpha1/volume_types.go` to test the new tasks:
- `./doit.sh format` should reformat the code according to `gofmt`
- `./doit.sh codegen` should re-generate the boilerplate code
- `./doit.sh lint:go` should complains if either the formatting is not good or the generated files are outdated.

---

Closes: #1410

### CODE REVIEW

**Comment by slaperche-scality** [comment]
Need to be sure it works on the CI (i.e.: green build) before removing the Draft label.
Also want to check if `dep` is really required, it may not be.

Other than that, you can start reviewing the code (lot of commit, but several of them are pretty small).

**Comment by NicolasT** [comment]
We don't (or shouldn't, at least) rely on `dep`, we're using Go modules and no vendoring.

File: buildchain/buildchain/config.py
Comment (NicolasT) [inline buildchain/buildchain/config.py:61]: `go fmt` ?

File: buildchain/buildchain/constants.py
Comment (NicolasT) [inline buildchain/buildchain/constants.py]: There should be no `vendor` directory. If there is one, that's a 'bug'.

File: eve/workers/pod-builder/Dockerfile
Comment (NicolasT) [inline eve/workers/pod-builder/Dockerfile]: I'd avoid using shorthands.

File: eve/workers/pod-builder/Dockerfile
Comment (NicolasT) [inline eve/workers/pod-builder/Dockerfile:44]: Consider checking the signature of the downloaded file?

File: buildchain/buildchain/config.py
Comment (slaperche-scality) [reply buildchain/buildchain/config.py:61]: I tried using `go fmt` but it only supports a subset of what `goftmt` does (no `-s` and more importantly, no `-d`)

```
% go help fmt
usage: go fmt [-n] [-x] [packages]

Fmt runs the command 'gofmt -l -w' on the packages named
by the import paths. It prints the names of the files that are modified.

For more about gofmt, see 'go doc cmd/gofmt'.
For more about specifying packages, see 'go help packages'.

The -n flag prints commands that would be executed.
The -x flag prints commands as they are executed.

To run gofmt with specific options, run gofmt itself.

See also: go fix, go vet.
% gofmt -h
usage: gofmt [flags] [path ...]
  -cpuprofile string
        write cpu profile to this file
  -d    display diffs instead of rewriting files
  -e    report all errors (not just the first 10 on different lines)
  -l    list files whose formatting differs from gofmt's
  -r string
        rewrite rule (e.g., 'a[b:len(a)] -> a[b:]')
  -s    simplify code
  -w    write result to (source) file instead of stdout
```
See also: https://stackoverflow.com/questions/55319495/what-is-the-difference-between-gofmt-go-fmt

Since both are available with a standard Go installation (`gofmt` isn't an extra-dependency), I went with `gofmt`.

File: buildchain/buildchain/constants.py
Comment (slaperche-scality) [reply buildchain/buildchain/constants.py]: Ho, you're right!!
The one I had is a vestigial directory of when I was using an old `operator-sdk`.

I guess we could also remove the entry from the `.gitignore` then?

File: eve/workers/pod-builder/Dockerfile
Comment (slaperche-scality) [reply eve/workers/pod-builder/Dockerfile:44]: :+1: 
I'll add it (anyway I have to modify this Dockerfile, things aren't working yet)

File: eve/workers/pod-builder/Dockerfile
Comment (slaperche-scality) [reply eve/workers/pod-builder/Dockerfile]: Done

File: eve/workers/pod-builder/Dockerfile
Comment (slaperche-scality) [reply eve/workers/pod-builder/Dockerfile:44]: That may be harder than expected, because I don't have the maintainers public key, so I have to download it but to get the ID I need to parse the error message from `gpg`

> To download the key, use the following command, replacing $KEY_ID with the RSA key string provided in the output of the previous command:

:/

File: buildchain/buildchain/constants.py
Comment (slaperche-scality) [reply buildchain/buildchain/constants.py]: Done

**Comment by slaperche-scality** [comment]
> We don't (or shouldn't, at least) rely on `dep`, we're using Go modules and no vendoring.

Yes, you're right: I got rid of it.

**Comment by slaperche-scality** [comment]
A green build \o/, I'm passing this as ready for review then :slightly_smiling_face: 

That being said, since it's adding 20 min to the CI time (because of the retries) we probably want to fix that.
But "it works" so it can be reviewed…

**Comment by NicolasT** [comment]
Adding 20min to CI is a no-go as far as I'm concerned. I believe there's a fairly easy fix though: we're running the build in a container, right? (If not, shouldn't we? ;-))

We can then apply the same 'trick' as we do for the doc building which downloads the Tox/Sphinx env, something along the lines of
```
WORKDIR /usr/src/storage-operator
COPY go.mod go.sum /usr/src/storage-operator/
RUN go mod download
COPY . /usr/src/storage-operator
RUN operator-sdk build ...
```

File: storage-operator/.gitignore
Comment (NicolasT) [inline storage-operator/.gitignore]: I wouldn't remove it, TBH. *Not* having it wouldn't prevent anyone from checking things into `vendor/`. Having it requires one to add `-f` to `git add`, otherwise Git won't be willing to touch files in that folder.

Comment (slaperche-scality) [reply storage-operator/.gitignore]: But how do they check things into `vendor`? The directory no longer exists when using 0.9.0.

**Comment by slaperche-scality** [comment]
@bert-e reset

File: eve/workers/pod-builder/Dockerfile
Comment (slaperche-scality) [reply eve/workers/pod-builder/Dockerfile:44]: Or should I hardcode the key when doing the `recv-keys`?

Comment (NicolasT) [reply eve/workers/pod-builder/Dockerfile:44]: Ugh, maybe file a TODO ticket but leave it for now.

File: eve/workers/pod-linter/Dockerfile
Comment (NicolasT) [inline eve/workers/pod-linter/Dockerfile:30]: Given we seem to be creating (almost) the same container image twice, does it really make sense to make a distinction between a `linter` and `builder` container?

File: storage-operator/.gitignore
Comment (NicolasT) [reply storage-operator/.gitignore]: `mkdir vendor`? `go mod vendor`?...

Boils down to: having `vendor/` in here doesn't do any harm, it might even provide a benefit, so I don't see any reason to remove it.

Comment (slaperche-scality) [reply storage-operator/.gitignore]: Fair enough, I'll add it back then.

But if we could have `vendor` here from `go mod vendor` I should keep the code to filter it then, no?

Comment (NicolasT) [reply storage-operator/.gitignore]: No, because we *shouldn't* have a `vendor` directory. This is a way to ensure we don't ever get one (unless one is very stubborn).

File: eve/workers/pod-builder/Dockerfile
Comment (slaperche-scality) [reply eve/workers/pod-builder/Dockerfile:44]: Tracked by #1444.

File: storage-operator/.gitignore
Comment (slaperche-scality) [reply storage-operator/.gitignore]: Ok seen with you directly: will keep the `.gitignore` entry but remove the code

File: eve/workers/pod-linter/Dockerfile
Comment (slaperche-scality) [reply eve/workers/pod-linter/Dockerfile:30]: I think the reason behind the split is that we want to use `centos:7` as base image (and that's what we do for the pod-builder).

Now, the issue with pod-linter is that we need a not too old `shellcheck`, which is not available on CentOS 7 so Jordi switched the base image to `fedora:29` for this pod (IIRC it was too cumbersome to get a recent `shellcheck` on CentOS)

That's the reason behind the split I think (keep using CentOS 7 everywhere, except for linting).

**Comment by slaperche-scality** [comment]
@bert-e reset

**Comment by slaperche-scality** [comment]
@bert-e reset

File: buildchain/buildchain/config.py
Comment (gdemonet) [inline buildchain/buildchain/config.py:66]: Mmh, works but could have been simpler (any command name containing dashes will need this special case). On the other hand, the heuristic is quite simple:

```python
def command_name(self):
    return self.name.lower().replace('_', '-')
```

File: buildchain/buildchain/image.py
Comment (gdemonet) [inline buildchain/buildchain/image.py:349]: Don't we want the MetalK8s version here?

File: eve/workers/pod-builder/Dockerfile
Comment (gdemonet) [inline eve/workers/pod-builder/Dockerfile]: Shouldn't these versions be set in `ARG` commands instead? Same as for `BUILDBOT_VERSION`

File: buildchain/buildchain/format.py
Comment (gdemonet) [inline buildchain/buildchain/format.py]: minor:

```suggestion
# List of available formatting tasks.
```

File: buildchain/buildchain/utils.py
Comment (gdemonet) [inline buildchain/buildchain/utils.py]: minor: Could be interesting to have these `title_with_<something>` wrapped in another function, taking only the `command` argument, since we'll always need a `function(task: types.Task) -> str` to pass in the task dict.

File: buildchain/buildchain/lint.py
Comment (gdemonet) [inline buildchain/buildchain/lint.py:128]: minor: for consistency with `check_go_fmt`

```suggestion
            msg='outdated generated Go files, please run `doit.sh codegen:go`'
```

File: buildchain/buildchain/lint.py
Comment (gdemonet) [inline buildchain/buildchain/lint.py]: This comment is missing an end I believe ;)

File: buildchain/buildchain/lint.py
Comment (gdemonet) [inline buildchain/buildchain/lint.py:126]: Should we reset the changes in that situation?

File: eve/workers/pod-builder/Dockerfile
Comment (gdemonet) [inline eve/workers/pod-builder/Dockerfile]: Since you're only `COPY`ing from `storage-operator/`, maybe the build context could be just `'storage-operator/'` instead of `'.'`?

**Review by gdemonet — COMMENTED** [review]
Few questions and minor comments, most of it looks good to me!

File: buildchain/buildchain/config.py
Comment (slaperche-scality) [reply buildchain/buildchain/config.py:66]: The problem is that some command does contains `_` in their name and that would break them.

True, we aren't using such command yet, but if one day we add one, I don't want the poor guy to go looking all over the buildchain trying to understand why the command he added doesn't work.

File: buildchain/buildchain/image.py
Comment (slaperche-scality) [reply buildchain/buildchain/image.py:349]: In the long term, we do want a real version number here (not latest).
For now: no, we will stay with latest.

Extract from a Slack discussion I had with Nicolas:

> Can use `latest` for now
> Until we figure out how to properly version `storage-operator`, also in its `version.go` file and such
> Can of worms

File: eve/workers/pod-builder/Dockerfile
Comment (slaperche-scality) [reply eve/workers/pod-builder/Dockerfile]: Done

File: eve/workers/pod-builder/Dockerfile
Comment (slaperche-scality) [reply eve/workers/pod-builder/Dockerfile]: Indeed. Done.

File: buildchain/buildchain/lint.py
Comment (slaperche-scality) [reply buildchain/buildchain/lint.py:126]: If we reset, we could lose the existing diff of the user.

The buildchain should avoid to mess with the git state as much as possible (yes, here we are modifying files, but anyway the guy is gonna run `./doit.sh codegen` so it's not a big deal).

File: buildchain/buildchain/lint.py
Comment (slaperche-scality) [reply buildchain/buildchain/lint.py:128]: Since the lint does run the codegen, the user don't really have to run it: we have done it for him already (would be better if `operator-sdk` had a dry-run mode, but that's not the case, so…)

File: buildchain/buildchain/format.py
Comment (slaperche-scality) [reply buildchain/buildchain/format.py]: Done (also fixed the same in `codegenpy`)

File: buildchain/buildchain/lint.py
Comment (slaperche-scality) [reply buildchain/buildchain/lint.py]: Indeed >_<

Done

File: buildchain/buildchain/utils.py
Comment (slaperche-scality) [reply buildchain/buildchain/utils.py]: Good idea!

Done in a dedicated commit.

**Comment by slaperche-scality** [comment]
@bert-e reset

**Comment by NicolasT** [comment]
It would make sense to hold off on this PR until #1447 is merged, then rebase it on `development/2.4` and change the target branch, so other version branches aren't impacted by the time spent on this in CI.

**Comment by slaperche-scality** [comment]
@bert-e reset

File: buildchain/buildchain/codegen.py
Comment (NicolasT) [inline buildchain/buildchain/codegen.py:27]: To be really picky, this should be `['k8s', 'openapi']` (since it's a `list<str>` rather than a `(str, str)`), but let's not let that keep this PR unmerged :-P

File: storage-operator/.dockerignore
Comment (NicolasT) [inline storage-operator/.dockerignore:1]: Not very fond we need this, but sure, for now. At some point I'd like to roll building of the whole thing in a container-based environment (indeed, maybe not using `operator-sdk build ...`), at which point this would need to be removed. Someday.

**Review by NicolasT — APPROVED** [review]

**Comment by slaperche-scality** [comment]
@bert-e approve

File: buildchain/buildchain/codegen.py
Comment (gdemonet) [reply buildchain/buildchain/codegen.py:27]: Why? Tuples are lighter than lists

**Comment by slaperche-scality** [comment]
@bert-e approve

File: storage-operator/.dockerignore
Comment (slaperche-scality) [reply storage-operator/.dockerignore:1]: Yes I added it to only upload the strict minimum to the docker daemon.

But that looks like premature optimization (and I even broke `operator-sdk build` when I added `build/` to the ignore => proves that it wasn't a great idea ^^").

I won't be against its removal when it becomes an hindrance :slightly_smiling_face:

File: buildchain/buildchain/codegen.py
Comment (slaperche-scality) [reply buildchain/buildchain/codegen.py:27]: I used tuple for its immutability (since the list is kind of a constant here), but in this context tuple and list are really alike: won't make any (observable) difference.

In a statically typed language, tuples and lists are a different beasts (lists are homogeneous whereas tuples aren't), but in Python where you can mix apples and oranges in both lists and tuples the key difference is the mutability.

Comment (NicolasT) [reply buildchain/buildchain/codegen.py:27]: The remark came indeed from a 'static typing' background:

- A *list* is an *ordered* collection of an *unknown* number of values of a *single* type
- A *tuple* is an *ordered* collection of a *known* number of values of (potentially) *different* (but known) types

As such, the type constructor of a list has a single argument, whilst the type constructor of an N-tuple has N arguments.

The immutability of a tuple doesn't really play a role in the argument.

Anyway, pedantic, though in general I'm not overly fond of conflating the two (yes, despite duck-typing).

**Pull request merged**
